src/audio/copier/copier.c

```diff
@@ -775,107 +775,89 @@ static int do_conversion_copy(struct comp_dev *dev,
-					break;
-
-				src_c = buffer_acquire(cd->endpoint_buffer[i]);
-				ret = do_conversion_copy(dev, cd, src_c, sink_c,
-							 &processed_data);
-				if (!ret) {
-					comp_update_buffer_consume(src_c,
-								   processed_data.source_bytes);
-					cd->output_total_data_processed +=
-						processed_data.sink_bytes;
-				}
-
-				buffer_release(src_c);
-
-				if (ret < 0)
-					break;
-			}
-
-			buffer_release(sink_c);
-		} else {
-			src = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
-			src_c = buffer_acquire(src);
-
-			for (i = 0; i < cd->endpoint_num; i++) {
-				sink_c = buffer_acquire(cd->endpoint_buffer[i]);
+	processed_data.source_bytes = 0;
 
-				ret = do_conversion_copy(dev, cd, src_c, sink_c,
-							 &processed_data);
-				buffer_release(sink_c);
-				if (ret < 0)
-					break;
-
-				ret = cd->endpoint[i]->drv->ops.copy(cd->endpoint[i]);
-				if (ret < 0)
-					break;
-			}
+	if (cd->endpoint_num >= 2) {
+		// TODO: Add implementation for multichannel ALH represented as multiple gateways
+		comp_err(dev, "Multichannel ALH (multiple gateways) support is NOT IMPLEMENTED");
+		return -1;
+	}
 
-			if (!ret) {
-				comp_update_buffer_consume(src_c,
-							   processed_data.source_bytes);
-				cd->input_total_data_processed += processed_data.source_bytes;
-			}
+	if (cd->endpoint_num && !cd->bsource_buffer) {
+		/* gateway as input */
+		ret = cd->endpoint[0]->drv->ops.copy(cd->endpoint[0]);
+		if (ret < 0)
+			return ret;
 
-			buffer_release(src_c);
-		}
+		src_c = buffer_acquire(cd->endpoint_buffer[0]);
 	} else {
-		/* do format conversion */
-		struct list_item *sink_list;
-		struct comp_copy_limits processed_data;
-
-		processed_data.source_bytes = 0;
-		processed_data.sink_bytes = 0;
-
+		/* component as input */
 		src = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
 		src_c = buffer_acquire(src);
 
-		/* do format conversion for each sink buffer */
-		list_for_item(sink_list, &dev->bsink_list) {
-			sink = container_of(sink_list, struct comp_buffer, source_list);
-			sink_c = buffer_acquire(sink);
-
+		if (cd->endpoint_num) {
+			/* gateway on output */
+			sink_c = buffer_acquire(cd->endpoint_buffer[0]);
 			ret = do_conversion_copy(dev, cd, src_c, sink_c, &processed_data);
 			buffer_release(sink_c);
+
 			if (ret < 0) {
-				comp_err(dev, "failed to copy buffer for comp %x",
-					 dev->ipc_config.id);
-				break;
+				buffer_release(src_c);
+				return ret;
+			}
+
+			ret = cd->endpoint[0]->drv->ops.copy(cd->endpoint[0]);
+			if (ret < 0) {
+				buffer_release(src_c);
+				return ret;
 			}
-			cd->output_total_data_processed += processed_data.sink_bytes;
 		}
+	}
 
-		if (!ret) {
-			comp_update_buffer_consume(src_c, processed_data.source_bytes);
-			cd->input_total_data_processed += processed_data.source_bytes;
+	/* zero or more components on outputs */
+	list_for_item(sink_list, &dev->bsink_list) {
+		sink = container_of(sink_list, struct comp_buffer, source_list);
```

Comment:
if gateway is on ouput when component is input, it has been copied in above case. so you need to skip this case in sink_list

Reply:
No, actually this case is one of the reasons for the fix. We should support the case of component on input, gateway on output pin 0 with up to 3 more components on outputs 1, 2, 3. This loop is to copy data to these additional components on outputs if present. I described the topologies we have to support in the comment above copier_copy().

Reply:
@serhiy-katsyuba-intel , if the gateway is the output, it is not in the sink list, right ? So no problem.

Reply:
Yes, gateway buffer is not on sink/source list. However, I do not know if this was specifically designed that way or just happened unintentionally and may change in future. So at least some comment that these buffers in a list are expected not to belong to a gateway might be helpful.
Luckily, this seems is verified on CI: every test that outputs audio would fail as it would produce twice as much data as expected (under a condition we use deep buffers).